src/mongo/jobs/DiscoverOutdatedComposites.class.php

```diff
@@ -0,0 +1,323 @@
+ *
+ * For a config with S specifications, each `perform()` call will therefore run S independent queries, each of which may
+ * return up to CURSOR_LIMIT documents to update, thus regenerating a maximum of `S*CURSOR_LIMIT` composite documents.
+ * 
+ * There are a number of things left TODO, as later considerations:
+ * 
+ * [1] Randomise the order of the specs, to avoid contention 
+ *     (e.g. one heavily-used view sucking up all regens in cursor)
+ *
+ * [2] Reduce number of queries. Currently, it fetches a collection for every composite spec, 
+ *     even when many composites will share the same collection (e.g. db.views on data1, or 
+ *     db.table_rows on data2). It should be possible to group by datasource/collection pair.
+ *     That would reduce the number of queries from |specifications| to |<datasource,collection>|
+ *
+ * @package Tripod\Mongo\Jobs
+ */
+class DiscoverOutdatedComposites extends JobBase {
+
+    const STORE_NAME_KEY = 'storeName';
+    const CURSOR_LIMIT = 'cursorLimit';
+
+    /**
+     * Run the DiscoverOutdatedComposites job
+     * @throws \Exception
+     */
+    public function perform()
+    {
+        $this->debugLog("DiscoverOutdatedComposites::perform() start");
+
+        $this->validateArgs();
+
+        $config = $this->args[self::TRIPOD_CONFIG_KEY];
+        $storeName = $this->args[self::STORE_NAME_KEY];
+        $cursorLimit = $this->args[self::CURSOR_LIMIT];
+
+        // set the config to what is received
+        \Tripod\Mongo\Config::setConfig($config);
+
+        // closure around $cursorLimit in $this->getRegenTaskForMetadata
+        $getRegenTaskForMetadata = function($metadatum) use ($cursorLimit) { 
+            return $this->getRegenTaskForMetadata($metadatum, $cursorLimit); 
+        };
+
+        // compile a list of all composite configuration metadata objects
+        $compositeMetadata = $this->getCompositeMetadata($config, $storeName);
+
+        // for each composite metadata instance, fetch regen tasks (where they exist), 
+        // which include specific CBDs to regenerate
+        $regenTasksIncludingNulls = array_map($getRegenTaskForMetadata, $compositeMetadata);
+        $regenTasks = array_filter($regenTasksIncludingNulls, 'isset');
+
+        // for collected regeneration tasks, run the update
+        // TODO: this will probably want to schedule updates, rather than running directly
+        $this->runRegenerationTasks($regenTasks);
+    }
+
+    /**
+     * @param array $data
+     * @param string|null $queueName
+     */
+    public function createJob(Array $data, $queueName=null)
+    {
+        if(!$queueName)
+        {
+            $queueName = Config::getDiscoverQueueName();
+        }
+        elseif(strpos($queueName, \Tripod\Mongo\Config::getDiscoverQueueName()) === false)
+        {
+            $queueName = \Tripod\Mongo\Config::getDiscoverQueueName() . '::' . $queueName;
```

Comment:
(but I will need to change the code further to achieve that - the current form uses the same queue, but that code will be replaced.)